### DISCUSSION TITLE
[C128] Relocate/fixed memory assign the functions within <stdio.h> <conio.h> <string.h>

### DISCUSSION THREAD

**Discussion by RetrofanC** [cc65/cc65#1739] [Q&A] [answered]
Hi Community, 

Would anybody has an idea what the best (most efficient, easiest, clever) way is to 'relocate/fixed assigned' the standard functions called by the <stdio.h> <conio.h> <string.h>? 

As I want them to become available in a particular segment of the memory! 

Just as a lucky guess I tried this which obviously didn't work; 

#pragma code-name (push, "COMMONRAM");
#include <stdio.h>
#pragma code-name (pop);

Will there be a way in which I can do this for the full header file or at least for all functions which are being called in my main C program from the header file? 

If, there is no better way available. I in that case the only option left for me to go to all the 'source/assembly'  '.s' files and add a; 
 .segment	"COMMONRAM"

In all of those? 
So they will be put that function/assembly code into that common RAM memory section?. 

That can be done of course but it's just a lot of work to rewrite/edit them all. 
Furthermore, I would like to keep the CC65 IDE as stock as possible. 

If anybody would have a clever idea that is appreciated! 
Thanks in advance,

**Comment by polluks** [comment]
Well, a header is a header, it does not contain any code at all.
You may change the linker config to use your COMMONRAM as default and put your own code in non-COMMONRAM.

**Reply by RetrofanC** [reply]
Thanks for your quick response and rather creative suggestion. 
That would work indeed for a PoC  and I surely will test it.
It might not be the right way for a permanent solution. 

So the additional question it triggers is this; 

**Do #pragma directives also work in header files?** 

So would it be possible to have something like this? 
Having only the printf function go into a different segment? 

======= example ======
<stdio.h>

............
void __fastcall__ perror (const char* s);

#pragma code-name (push, "COMMONRAM");

int printf (const char* format, ...);

#pragma code-name (pop);
...........

================

Or would the reply be similar as before perhaps? 
" Well, a function declaration is a function declaration, it does not contain any code at all." 

So a #pragma only works when it is wrapped around the actual function definition and code? 

Thanks!

**Comment by mrdudz** [comment, marked as answer]
> So a #pragma only works when it is wrapped around the actual function definition and code?

In this case and for what you want to do - yes. The compiler must see that pragma when it is compiling the code, the segment is saved in the object file.

That said, somehow i vaguely recall there was a way to modify the object file - but i dont recall how (or i might even mix it up with a different compiler i used before). That would be the most elegant solution anyway :)

**Reply by RetrofanC** [reply]
Understood! 
Thanks for helping and clarify on this!
